integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+        case TMRESUME:
+            return "TMRESUME (" + flags + ")";
+        case TMSTARTRSCAN:
+            return "TMSTARTRSCAN (" + flags + ")";
+        case TMSUCCESS:
+            return "TMSUCCESS (" + flags + ")";
+        case TMSUSPEND:
+            return "TMSUSPEND (" + flags + ")";
+        default:
+            return String.valueOf(flags);
+        }
+    }
+
+
+    /*
+     * Inner and nested classes.
+     */
+
+
+    static record Association(BranchState branchState,
+                              Xid xid,
+                              boolean suspended,
+                              Connection connection,
+                              boolean priorAutoCommit) {
+
+        private static final Logger LOGGER = Logger.getLogger(Association.class.getName());
+
+        // Branch Association States: (XA specification, table 6-2)
+        // T0: Not Associated
+        // T1: Associated
+        // T2: Association Suspended
+
+        // Branch States: (XA specification, table 6-4)
+        // S0: Non-existent Transaction
+        // S1: Active
+        // S2: Idle
+        // S3: Prepared
+        // S4: Rollback Only
+        // S5: Heuristically Completed
+
+        Association(BranchState branchState, Xid xid, Connection connection) {
+            this(branchState, xid, false, connection);
+        }
+
+        Association(BranchState branchState, Xid xid, boolean suspended, Connection connection) {
+            this(branchState, xid, suspended, connection, true /* JDBC default; will be set from connection anyway */);
+        }
+
+        Association {
+            Objects.requireNonNull(xid, "xid");
+            switch (branchState) {
+            case IDLE:
+                break;
+            case ACTIVE:
+            case HEURISTICALLY_COMPLETED:
+            case NON_EXISTENT_TRANSACTION:
+            case PREPARED:
+            case ROLLBACK_ONLY:
+                if (suspended) {
+                    throw new IllegalArgumentException("suspended");
+                }
+                break;
+            default:
+                throw new IllegalArgumentException("branchState: " + branchState);
+            }
+            try {
+                priorAutoCommit = connection.getAutoCommit();
+                if (priorAutoCommit) {
+                    connection.setAutoCommit(false);
+                }
+            } catch (SQLException sqlException) {
+                throw new UncheckedSQLException(sqlException);
+            }
+            // T0, T1 or T2; S0 or S2
+        }
+
+        public boolean suspended() {
+            assert this.suspended ? this.branchState() == BranchState.IDLE : true;
+            return this.suspended;
+        }
+
+        private Association activeToIdle() {
+            if (!this.suspended()) {
+                switch (this.branchState()) {
+                case ACTIVE:
+                    // OK; end(*) was called and didn't fail with an XAER_RB* code and we are not suspended
```

Comment:
You'll see notation like `end(*)`: that's shorthand for `end(<any flags>)`, i.e. `end` was called with any possible legal flag, i.e. `TMSUCCESS` or what have you.